_posts/wasmer-go-embedding-1.0.md

```diff
@@ -0,0 +1,534 @@
+cpuFeatures := wasmer.NewCpuFeatures()
+cpuFeatures.Add("sse2")
+
+// Finally, a Target is a pair composed of a Triple and CPU features.
+target := wasmer.NewTarget(triple, cpuFeatures)
+```
+
+After that, we pass the `Target` to the `Config` API, we compile the
+WebAssembly module, and save it:
+
+```go
+// Compile for a specific target.
+config := wasmer.NewConfig().UseTarget(target)
+
+// Create the engine with a specific configuration.
+engine := wasmer.NewEngineWithConfig(config)
+store := wasmer.NewStore(engine)
+
+// Let's compile the module for the given target.
+module, _ := wasmer.NewModule(store, wasmBytes)
+
+// Serialize the compiled module.
+ioutil.Writefile("my_wsam_module.wjit", module.Serialize(), 0644)
+```
+
+The serialized compiled module can be then deserialized on the
+targeted machine, instantiated, and executed!
+
+That way, it is easier to pre-compile any WebAssembly modules for a
+variety of machines!
+
+## WASI
+
+The `wasmer-go` package now features
+[WASI](https://github.com/WebAssembly/wasi) (The WebAssembly System
+Interface), with all the snapshot previews; understand, all the
+versions.
+
+First, the `GetWasiVersion` function can be used to know which version
+of WASI a WebAssembly module is using or whether it's not using WASI
+at all:
+
+```go
+module, _ := wasmer.NewModule(store, wasmBytes)
+
+fmt.Println(GetWasiVersion(module))
+```
+
+Second, to setup WASI, it's required to start by creating a
+`WasiEnvironment` with the help of the `NewWasiStateBuilder` API. The
+generated `WasiEnvironment` object is then used to generate an
+`ImportObject`. This later contains all the imports that “bridge” the
+WebAssembly module to the host to make WASI a reality. Of course, it's
+possible to use this `ImportObject` to import your own host functions,
+memories etc., just like with any `ImportObject`. `WasiEnvironment` is
+also responsible to read the `stdout` and `stderr` streams if they are
+captured.
+
+Let's see with an example. We want to execute [this Rust
+program](https://github.com/wasmerio/wasmer-go/blob/794da4bc5618f7e5f8e6f046d765033372763824/wasmer/testdata/wasi.rs),
+that prints its arguments, its environment variables, and that lists
+the content of its current working directory.
+
+```go
+engine := wasmer.NewEngine()
+store := wasmer.NewStore(engine)
+module, _ := wasmer.NewModule(store, wasmBytes)
+
+// We specify the program name: `test-program`. We also specify the
+// program is invoked with the `--test` argument, in addition to two
+// environment variables: `COLOR` and `APP_SHOULD_LOG`. Finally, we map
+// the `the_host_current_directory` to the current directory.
+wasiEnv, _ := wasmer.NewWasiStateBuilder("wasi-test-program").
+	Argument("--test")
+	Environment("COLOR", "true")
+	Environment("APP_SHOULD_LOG", "false")
+	MapDirectory("the_host_current_directory", ".").
+	CaptureStdout()
+	Finalize()
+
+// Get the import object!
+importObject, _ := wasiEnv.GenerateImportObject(store, module)
```

Comment:
Not sure if you mentioned that this will auto-detect the version or not. Not sure if it's worth bringing up

Reply:
Initially, I was talking about it, but I found it was adding too much information.

Reply:
I dropped a line:

```go
// Get the import object (WASI version is auto-detected)!
importObject, _ := wasiEnv.GenerateImportObject(store, module)
```